base_layer/wallet/src/storage/sqlite_db/wallet.rs

```diff
@@ -533,39 +581,24 @@ fn get_cipher_for_db_encryption(
             let mut rng = OsRng;
             rng.fill_bytes(main_key.reveal_mut());
 
-            // We'll be encrypting the main key shortly, so keep a clone around
-            let main_key_clone = main_key.clone();
-
             // Use the most recent `Argon2` parameters
             let argon2_params = Argon2Parameters::from_version(None)?;
 
             // Derive the secondary key from the user's passphrase and a high-entropy salt
             let secondary_key_salt = SaltString::generate(&mut rng);
-            let mut secondary_key = WalletSecondaryEncryptionKey::from(SafeArray::default());
-            argon2::Argon2::new(argon2_params.algorithm, argon2_params.version, argon2_params.params)
-                .hash_password_into(
-                    passphrase.reveal(),
-                    secondary_key_salt.as_bytes(),
-                    secondary_key.reveal_mut(),
-                )
-                .map_err(|e| WalletStorageError::AeadError(e.to_string()))?;
-
-            // Use the secondary key to encrypt the main key, authenticating with the version to mitigate mismatch
-            // attacks
-            let main_key_cipher = XChaCha20Poly1305::new(Key::from_slice(secondary_key.reveal()));
-            let mut aad = main_key_domain;
-            aad.push(argon2_params.id);
-            let encrypted_main_key =
-                encrypt_bytes_integral_nonce(&main_key_cipher, aad, Hidden::hide(main_key.reveal().clone()))
-                    .map_err(WalletStorageError::AeadError)?;
+            let secondary_key =
+                derive_secondary_key(passphrase, argon2_params.clone(), &secondary_key_salt.to_string())?;
+
+            // Use the secondary key to encrypt the main key
+            let encrypted_main_key = encrypt_main_key(&secondary_key, &main_key, argon2_params.id)?;
 
             // Store the secondary key version, secondary key salt, and encrypted main key
-            WalletSettingSql::new(DbKey::SecondaryKeyVersion, argon2_params.id.to_string()).set(conn)?;
-            WalletSettingSql::new(DbKey::SecondaryKeySalt, secondary_key_salt.to_string()).set(conn)?;
-            WalletSettingSql::new(DbKey::EncryptedMainKey, encrypted_main_key.to_hex()).set(conn)?;
+            WalletSettingSql::new(DbKey::SecondaryKeyVersion, argon2_params.id.to_string()).set(&conn)?;
+            WalletSettingSql::new(DbKey::SecondaryKeySalt, secondary_key_salt.to_string()).set(&conn)?;
+            WalletSettingSql::new(DbKey::EncryptedMainKey, encrypted_main_key.to_hex()).set(&conn)?;
```

Comment:
Is this necessary? If any of the operations were to fail, no encryption would be set up, the function would return an error, and no data would be corrupted.

Reply:
I don't think is particularly necessary, since a failure in one of these operations should not result in any data loss. But it doesn't hurt, so it's been added.